### ISSUE TITLE
`global_train_batch_size` doesn't affect the number of steps, bug in gradient accumulation?

### ISSUE THREAD

**Issue by dtamayo-nlp** [AnswerDotAI/ModernBERT#247]
Hi, 

I am encountering that changing the `global_train_batch_size` variable doesn't affect the number of steps required to increase the lr rate, which shouldn't be the case. If we duplicate the number of tokens on each update, and we keep the number of tokens in the learning rate warmup as `3_000_000_000tok` as provided in your pre-training scripts, the number of steps required should decrease by 1/2. However, this is not the case:

<img width="1394" height="713" alt="Image" src="https://github.com/user-attachments/assets/823365cc-e2e1-4945-9747-282c0e370010" />

On the other hand, when launching with different number of GPUs/nodes, let's say we increase the number of GPUs by a factor xN, I see this slope change by a factor N, I am trying to debug the point in which we execute the step of the optimizer, but I don't see an accumulation of batches happening in my case ([code](https://github.com/mosaicml/composer/blob/4ae29b1afec56ce2d54f6fa07a7f9578a0d364b0/composer/trainer/trainer.py#L2660-L2777)).

I am probably doing something wrong, but am I the only one experimenting this issue?

**Comment by dtamayo-nlp** [comment]
My mistake, this is related to the fact that in the yaml there's a warmup in the batch size used. It is also mentioned in the paper.
